lib/api/normalize.js

```diff
@@ -0,0 +1,307 @@
+            ENTITY_TYPE.EMBED,
+        ]);
+
+        nextEditorState = this.resetBlockDepth(nextEditorState, maxListNesting);
+
+        nextEditorState = this.resetBlockType(
+            nextEditorState,
+            enabledBlockTypes,
+        );
+
+        nextEditorState = this.filterInlineStyle(
+            nextEditorState,
+            enabledInlineStyles,
+        );
+
+        nextEditorState = this.resetAtomicBlocks(
+            nextEditorState,
+            enabledEntityTypes,
+        );
+
+        nextEditorState = this.filterEntityType(
+            nextEditorState,
+            enabledEntityTypes,
+        );
+
+        return nextEditorState;
+    },
+
+    /**
+     * Makes atomic blocks where they would be required for a block-level entity
+     * to work correctly, when such an entity exists.
+     * Note: at the moment, this is only useful for IMAGE entities that Draft.js
+     * injects on arbitrary blocks on paste.
+     */
+    preserveAtomicBlocks(editorState, entityTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const perservedBlocks = blockMap
+            .filter(block => {
+                const entityKey = block.getEntityAt(0);
+
+                return (
+                    entityKey &&
+                    entityTypes.includes(content.getEntity(entityKey).getType())
+                );
+            })
+            .map(block => block.set('type', BLOCK_TYPE.ATOMIC));
+
+        if (perservedBlocks.size !== 0) {
+            return EditorState.set(editorState, {
+                currentContent: content.merge({
+                    blockMap: blockMap.merge(perservedBlocks),
+                }),
+            });
+        }
+
+        return editorState;
+    },
+
+    /**
+     * Resets the depth of all the content to at most maxListNesting.
+     */
+    resetBlockDepth(editorState, maxListNesting) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const changedBlocks = blockMap
+            .filter(block => block.getDepth() > maxListNesting)
+            .map(block => block.set('depth', maxListNesting));
+
+        if (changedBlocks.size !== 0) {
+            return EditorState.set(editorState, {
+                currentContent: content.merge({
+                    blockMap: blockMap.merge(changedBlocks),
+                }),
+            });
+        }
+
+        return editorState;
+    },
+
+    /**
+     * Resets all blocks that use unavailable types to unstyled.
+     */
+    resetBlockType(editorState, enabledTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const changedBlocks = blockMap
+            .filter(block => !enabledTypes.includes(block.getType()))
+            .map(block => block.set('type', BLOCK_TYPE.UNSTYLED));
+
+        if (changedBlocks.size !== 0) {
```

Comment:
Note that if `UNSTYLED` isn't in the enabled types, a change will be created from unstyled to unstyled. Seems fine though.